File: sdk/python/urlshortener/services/webhooks.py
"""Webhook service for URL Shortener API."""

from typing import List, Optional, Dict, Any

from ..http_client import HTTPClient, AsyncHTTPClient
from ..types.webhooks import (
    Webhook,
    CreateWebhookRequest,
    UpdateWebhookRequest,
    WebhookDelivery,
    WebhookStats,
    WebhookPayload,
    WebhookEventCategories,
    WebhookTestResult,
    WebhookHealthSummary,
    WebhookTemplate,
    WebhookInsights,
    WebhookFilter,
    WebhookDeliveryFilter,
    WebhookEvent,
    WebhookStatus
)
from ..types.common import RequestConfig, PaginationParams, PaginatedResponse


class WebhookService:
    """Synchronous webhook service."""

    def __init__(self, http_client: HTTPClient):
        """Initialize webhook service.
        
        Args:
            http_client: HTTP client instance
        """
        self.http_client = http_client

    def create(
        self,
        name: str,
        url: str,
        events: List[WebhookEvent],
        secret: Optional[str] = None,
        max_retries: Optional[int] = None,
        timeout_seconds: Optional[int] = None,
        config: Optional[RequestConfig] = None
    ) -> Webhook:
        """Create a new webhook.
        
        Args:
            name: Webhook name
            url: Webhook endpoint URL
            events: Events to subscribe to
            secret: Webhook secret for verification
            max_retries: Maximum retry attempts
            timeout_seconds: Request timeout
            config: Request configuration
            
        Returns:
            Webhook: Created webhook
        """
        request = CreateWebhookRequest(
            name=name,
            url=url,
            events=events,
            secret=secret,
            max_retries=max_retries,
            timeout_seconds=timeout_seconds
        )
        
        response = self.http_client.post(
            '/webhooks',
            data=request.dict(exclude_none=True),
            config=config
        )
        
        return Webhook(**response['data'])

    def get(self, webhook_id: int, config: Optional[RequestConfig] = None) -> Webhook:
        """Get a webhook by ID.
        
        Args:
            webhook_id: Webhook ID
            config: Request configuration
            
        Returns:
            Webhook: Webhook details
        """
        response = self.http_client.get(f'/webhooks/{webhook_id}', config=config)
        return Webhook(**response['data'])

    def list(
        self,
        pagination: Optional[PaginationParams] = None,
        filter_params: Optional[WebhookFilter] = None,
        config: Optional[RequestConfig] = None
    ) -> PaginatedResponse[Webhook]:
        """List user's webhooks.
        
        Args:
            pagination: Pagination parameters
            filter_params: Filter parameters
            config: Request configuration
            
        Returns:
            PaginatedResponse[Webhook]: Paginated list of webhooks
        """
        params = {}
        
        if pagination:
            params.update(pagination.dict(exclude_none=True))
            
        if filter_params:
            params.update(filter_params.dict(exclude_none=True))
            
        response = self.http_client.get('/webhooks', params=params, config=config)
        
        webhooks = [Webhook(**item) for item in response['data']]
        return PaginatedResponse[Webhook](
            data=webhooks,
            total=response['total'],
            limit=response['limit'],
            offset=response['offset']
        )

    def update(
        self,
        webhook_id: int,
        name: Optional[str] = None,
        url: Optional[str] = None,
        events: Optional[List[WebhookEvent]] = None,
        secret: Optional[str] = None,
        status: Optional[WebhookStatus] = None,
        max_retries: Optional[int] = None,
        timeout_seconds: Optional[int] = None,
        config: Optional[RequestConfig] = None
    ) -> Webhook:
        """Update a webhook.
        
        Args:
            webhook_id: Webhook ID
            name: Webhook name
            url: Webhook endpoint URL
            events: Events to subscribe to
            secret: Webhook secret
            status: Webhook status
            max_retries: Maximum retry attempts
            timeout_seconds: Request timeout
            config: Request configuration
            
        Returns:
            Webhook: Updated webhook
        """
        request = UpdateWebhookRequest(
            name=name,
            url=url,
            events=events,
            secret=secret,
            status=status,
            max_retries=max_retries,
            timeout_seconds=timeout_seconds
        )
        
        response = self.http_client.put(
            f'/webhooks/{webhook_id}',
            data=request.dict(exclude_none=True),
            config=config
        )
        
        return Webhook(**response['data'])

    def delete(self, webhook_id: int, config: Optional[RequestConfig] = None) -> bool:
        """Delete a webhook.
        
        Args:
            webhook_id: Webhook ID
            config: Request configuration
            
        Returns:
            bool: True if deleted successfully
        """
        response = self.http_client.delete(f'/webhooks/{webhook_id}', config=config)
        return response.get('success', False)

    def test(
        self,
        webhook_id: int,
        event_type: Optional[WebhookEvent] = None,
        config: Optional[RequestConfig] = None
    ) -> WebhookTestResult:
        """Test a webhook endpoint.
        
        Args:
            webhook_id: Webhook ID
            event_type: Event type to test (uses default if not specified)
            config: Request configuration
            
        Returns:
            WebhookTestResult: Test result
        """
        data = {}
        if event_type:
            data['event_type'] = event_type
            
        response = self.http_client.post(
            f'/webhooks/{webhook_id}/test',
            data=data,
            config=config
        )
        
        return WebhookTestResult(**response['data'])

    def get_deliveries(
        self,
        webhook_id: int,
        pagination: Optional[PaginationParams] = None,
        filter_params: Optional[WebhookDeliveryFilter] = None,
        config: Optional[RequestConfig] = None
    ) -> PaginatedResponse[WebhookDelivery]:
        """Get webhook delivery history.
        
        Args:
            webhook_id: Webhook ID
            pagination: Pagination parameters
            filter_params: Filter parameters
            config: Request configuration
            
        Returns:
            PaginatedResponse[WebhookDelivery]: Paginated delivery history
        """
        params = {}
        
        if pagination:
            params.update(pagination.dict(exclude_none=True))
            
        if filter_params:
            params.update(filter_params.dict(exclude_none=True))
            
        response = self.http_client.get(
            f'/webhooks/{webhook_id}/deliveries',
            params=params,
            config=config
        )
        
        deliveries = [WebhookDelivery(**item) for item in response['data']]
        return PaginatedResponse[WebhookDelivery](
            data=deliveries,
            total=response['total'],
            limit=response['limit'],
            offset=response['offset']
        )

    def get_delivery(
        self,
        webhook_id: int,
        delivery_id: int,
        config: Optional[RequestConfig] = None
    ) -> WebhookDelivery:
        """Get a specific webhook delivery.
        
        Args:
            webhook_id: Webhook ID
            delivery_id: Delivery ID
            config: Request configuration
            
        Returns:
            WebhookDelivery: Delivery details
        """
        response = self.http_client.get(
            f'/webhooks/{webhook_id}/deliveries/{delivery_id}',
            config=config
        )
        
        return WebhookDelivery(**response['data'])

    def retry_delivery(
        self,
        webhook_id: int,
        delivery_id: int,
        config: Optional[RequestConfig] = None
    ) -> WebhookDelivery:
        """Retry a failed webhook delivery.
        
        Args:
            webhook_id: Webhook ID
            delivery_id: Delivery ID
            config: Request configuration
            
        Returns:
            WebhookDelivery: Retry result
        """
        response = self.http_client.post(
            f'/webhooks/{webhook_id}/deliveries/{delivery_id}/retry',
            config=config
        )
        
        return WebhookDelivery(**response['data'])

    def get_stats(
        self,
        webhook_id: int,
        period: Optional[str] = None,
        config: Optional[RequestConfig] = None
    ) -> WebhookStats:
        """Get webhook delivery statistics.
        
        Args:
            webhook_id: Webhook ID
            period: Time period for statistics
            config: Request configuration
            
        Returns:
            WebhookStats: Delivery statistics
        """
        params = {}
        if period:
            params['period'] = period
            
        response = self.http_client.get(
            f'/webhooks/{webhook_id}/stats',
            params=params,
            config=config
        )
        
        return WebhookStats(**response['data'])

    def get_event_categories(
        self,
        config: Optional[RequestConfig] = None
    ) -> WebhookEventCategories:
        """Get available webhook event categories.
        
        Args:
            config: Request configuration
            
        Returns:
            WebhookEventCategories: Available event categories
        """
        response = self.http_client.get('/webhooks/events', config=config)
        return WebhookEventCategories(**response['data'])

    def get_health_summary(
        self,
        config: Optional[RequestConfig] = None
    ) -> WebhookHealthSummary:
        """Get webhook health summary.
        
        Args:
            config: Request configuration
            
        Returns:
            WebhookHealthSummary: Health summary
        """
        response = self.http_client.get('/webhooks/health', config=config)
        return WebhookHealthSummary(**response['data'])

    def get_templates(
        self,
        config: Optional[RequestConfig] = None
    ) -> List[WebhookTemplate]:
        """Get webhook templates for common use cases.
        
        Args:
            config: Request configuration
            
        Returns:
            List[WebhookTemplate]: Available templates
        """
        response = self.http_client.get('/webhooks/templates', config=config)
        return [WebhookTemplate(**item) for item in response['data']]

    def get_insights(
        self,
        webhook_id: Optional[int] = None,
        period: Optional[str] = None,
        config: Optional[RequestConfig] = None
    ) -> WebhookInsights:
        """Get webhook insights and recommendations.
        
        Args:
            webhook_id: Specific webhook ID (or all webhooks if not specified)
            period: Time period for analysis
            config: Request configuration
            
        Returns:
            WebhookInsights: Insights and recommendations
        """
        params = {}
        if webhook_id:
            params['webhook_id'] = webhook_id
        if period:
            params['period'] = period
            
        response = self.http_client.get(
            '/webhooks/insights',
            params=params,
            config=config
        )
        
        return WebhookInsights(**response['data'])

    def pause(self, webhook_id: int, config: Optional[RequestConfig] = None) -> Webhook:
        """Pause a webhook (set status to inactive).
        
        Args:
            webhook_id: Webhook ID
            config: Request configuration
            
        Returns:
            Webhook: Updated webhook
        """
        response = self.http_client.post(f'/webhooks/{webhook_id}/pause', config=config)
        return Webhook(**response['data'])

    def resume(self, webhook_id: int, config: Optional[RequestConfig] = None) -> Webhook:
        """Resume a webhook (set status to active).
        
        Args:
            webhook_id: Webhook ID
            config: Request configuration
            
        Returns:
            Webhook: Updated webhook
        """
        response = self.http_client.post(f'/webhooks/{webhook_id}/resume', config=config)
        return Webhook(**response['data'])

    def verify_signature(
        self,
        payload: str,
        signature: str,
        secret: str
    ) -> bool:
        """Verify webhook signature locally.
        
        Args:
            payload: Raw webhook payload
            signature: Webhook signature header
            secret: Webhook secret
            
        Returns:
            bool: True if signature is valid
        """
        import hmac
        import hashlib
        
        # Compute expected signature
        expected_signature = hmac.new(
            secret.encode('utf-8'),
            payload.encode('utf-8'),
            hashlib.sha256
        ).hexdigest()
        
        # Compare signatures
        expected_signature = f"sha256={expected_signature}"
        return hmac.compare_digest(expected_signature, signature)


class AsyncWebhookService:
    """Asynchronous webhook service."""

    def __init__(self, http_client: AsyncHTTPClient):
        """Initialize async webhook service.
        
        Args:
            http_client: Async HTTP client instance
        """
        self.http_client = http_client

    async def create(
        self,
        name: str,
        url: str,
        events: List[WebhookEvent],
        secret: Optional[str] = None,
        max_retries: Optional[int] = None,
        timeout_seconds: Optional[int] = None,
        config: Optional[RequestConfig] = None
    ) -> Webhook:
        """Create a new webhook.
        
        Args:
            name: Webhook name
            url: Webhook endpoint URL
            events: Events to subscribe to
            secret: Webhook secret for verification
            max_retries: Maximum retry attempts
            timeout_seconds: Request timeout
            config: Request configuration
            
        Returns:
            Webhook: Created webhook
        """
        request = CreateWebhookRequest(
            name=name,
            url=url,
            events=events,
            secret=secret,
            max_retries=max_retries,
            timeout_seconds=timeout_seconds
        )
        
        response = await self.http_client.post(
            '/webhooks',
            data=request.dict(exclude_none=True),
            config=config
        )
        
        return Webhook(**response['data'])

    async def get(self, webhook_id: int, config: Optional[RequestConfig] = None) -> Webhook:
        """Get a webhook by ID.
        
        Args:
            webhook_id: Webhook ID
            config: Request configuration
            
        Returns:
            Webhook: Webhook details
        """
        response = await self.http_client.get(f'/webhooks/{webhook_id}', config=config)
        return Webhook(**response['data'])

    async def list(
        self,
        pagination: Optional[PaginationParams] = None,
        filter_params: Optional[WebhookFilter] = None,
        config: Optional[RequestConfig] = None
    ) -> PaginatedResponse[Webhook]:
        """List user's webhooks.
        
        Args:
            pagination: Pagination parameters
            filter_params: Filter parameters
            config: Request configuration
            
        Returns:
            PaginatedResponse[Webhook]: Paginated list of webhooks
        """
        params = {}
        
        if pagination:
            params.update(pagination.dict(exclude_none=True))
            
        if filter_params:
            params.update(filter_params.dict(exclude_none=True))
            
        response = await self.http_client.get('/webhooks', params=params, config=config)
        
        webhooks = [Webhook(**item) for item in response['data']]
        return PaginatedResponse[Webhook](
            data=webhooks,
            total=response['total'],
            limit=response['limit'],
            offset=response['offset']
        )

    async def update(
        self,
        webhook_id: int,
        name: Optional[str] = None,
        url: Optional[str] = None,
        events: Optional[List[WebhookEvent]] = None,
        secret: Optional[str] = None,
        status: Optional[WebhookStatus] = None,
        max_retries: Optional[int] = None,
        timeout_seconds: Optional[int] = None,
        config: Optional[RequestConfig] = None
    ) -> Webhook:
        """Update a webhook.
        
        Args:
            webhook_id: Webhook ID
            name: Webhook name
            url: Webhook endpoint URL
            events: Events to subscribe to
            secret: Webhook secret
            status: Webhook status
            max_retries: Maximum retry attempts
            timeout_seconds: Request timeout
            config: Request configuration
            
        Returns:
            Webhook: Updated webhook
        """
        request = UpdateWebhookRequest(
            name=name,
            url=url,
            events=events,
            secret=secret,
            status=status,
            max_retries=max_retries,
            timeout_seconds=timeout_seconds
        )
        
        response = await self.http_client.put(
            f'/webhooks/{webhook_id}',
            data=request.dict(exclude_none=True),
            config=config
        )
        
        return Webhook(**response['data'])

    async def delete(self, webhook_id: int, config: Optional[RequestConfig] = None) -> bool:
        """Delete a webhook.
        
        Args:
            webhook_id: Webhook ID
            config: Request configuration
            
        Returns:
            bool: True if deleted successfully
        """
        response = await self.http_client.delete(f'/webhooks/{webhook_id}', config=config)
        return response.get('success', False)

    async def test(
        self,
        webhook_id: int,
        event_type: Optional[WebhookEvent] = None,
        config: Optional[RequestConfig] = None
    ) -> WebhookTestResult:
        """Test a webhook endpoint.
        
        Args:
            webhook_id: Webhook ID
            event_type: Event type to test (uses default if not specified)
            config: Request configuration
            
        Returns:
            WebhookTestResult: Test result
        """
        data = {}
        if event_type:
            data['event_type'] = event_type
            
        response = await self.http_client.post(
            f'/webhooks/{webhook_id}/test',
            data=data,
            config=config
        )
        
        return WebhookTestResult(**response['data'])

    async def get_stats(
        self,
        webhook_id: int,
        period: Optional[str] = None,
        config: Optional[RequestConfig] = None
    ) -> WebhookStats:
        """Get webhook delivery statistics.
        
        Args:
            webhook_id: Webhook ID
            period: Time period for statistics
            config: Request configuration
            
        Returns:
            WebhookStats: Delivery statistics
        """
        params = {}
        if period:
            params['period'] = period
            
        response = await self.http_client.get(
            f'/webhooks/{webhook_id}/stats',
            params=params,
            config=config
        )
        
        return WebhookStats(**response['data'])

    def verify_signature(
        self,
        payload: str,
        signature: str,
        secret: str
    ) -> bool:
        """Verify webhook signature locally.
        
        Args:
            payload: Raw webhook payload
            signature: Webhook signature header
            secret: Webhook secret
            
        Returns:
            bool: True if signature is valid
        """
        import hmac
        import hashlib
        
        # Compute expected signature
        expected_signature = hmac.new(
            secret.encode('utf-8'),
            payload.encode('utf-8'),
            hashlib.sha256
        ).hexdigest()
        
        # Compare signatures
        expected_signature = f"sha256={expected_signature}"
        return hmac.compare_digest(expected_signature, signature)

File: sdk/javascript/src/services/webhooks.ts
import { HTTPClient } from '../http-client';
import { API_ENDPOINTS } from '../config';
import { 
  Webhook,
  WebhookEvent,
  CreateWebhookRequest,
  UpdateWebhookRequest,
  WebhookDelivery,
  WebhookStats,
  PaginationParams,
  PaginatedResponse,
  RequestConfig 
} from '../types';

export interface WebhookEventCategories {
  url_events: WebhookEvent[];
  analytics_events: WebhookEvent[];
  user_events: WebhookEvent[];
  system_events: WebhookEvent[];
}

export interface WebhookTestResult {
  delivery: WebhookDelivery;
  success: boolean;
  responseTime: number;
  statusCode: number;
  errorMessage?: string;
}

export class WebhookService {
  constructor(private client: HTTPClient) {}

  /**
   * Create a new webhook
   */
  async create(request: CreateWebhookRequest, config?: RequestConfig): Promise<Webhook> {
    return this.client.post<Webhook>(API_ENDPOINTS.WEBHOOKS, request, config);
  }

  /**
   * Get all webhooks for the authenticated user
   */
  async list(
    pagination?: PaginationParams,
    config?: RequestConfig
  ): Promise<PaginatedResponse<Webhook>> {
    const params = new URLSearchParams();
    
    if (pagination?.limit) {
      params.set('limit', pagination.limit.toString());
    }
    if (pagination?.offset) {
      params.set('offset', pagination.offset.toString());
    }

    const url = `${API_ENDPOINTS.WEBHOOKS}${params.toString() ? `?${params.toString()}` : ''}`;
    const response = await this.client.get<{
      webhooks: Webhook[];
      total: number;
      limit: number;
      offset: number;
    }>(url, config);

    return {
      data: response.webhooks,
      total: response.total,
      limit: response.limit,
      offset: response.offset,
    };
  }

  /**
   * Get a specific webhook by ID
   */
  async getById(id: number, config?: RequestConfig): Promise<Webhook> {
    return this.client.get<Webhook>(API_ENDPOINTS.WEBHOOK_BY_ID(id), config);
  }

  /**
   * Update an existing webhook
   */
  async update(id: number, request: UpdateWebhookRequest, config?: RequestConfig): Promise<Webhook> {
    return this.client.put<Webhook>(API_ENDPOINTS.WEBHOOK_BY_ID(id), request, config);
  }

  /**
   * Delete a webhook
   */
  async delete(id: number, config?: RequestConfig): Promise<void> {
    await this.client.delete(API_ENDPOINTS.WEBHOOK_BY_ID(id), config);
  }

  /**
   * Activate a webhook
   */
  async activate(id: number, config?: RequestConfig): Promise<void> {
    await this.client.post(API_ENDPOINTS.WEBHOOK_ACTIVATE(id), {}, config);
  }

  /**
   * Deactivate a webhook
   */
  async deactivate(id: number, config?: RequestConfig): Promise<void> {
    await this.client.post(API_ENDPOINTS.WEBHOOK_DEACTIVATE(id), {}, config);
  }

  /**
   * Test a webhook with a sample payload
   */
  async test(id: number, config?: RequestConfig): Promise<WebhookTestResult> {
    const delivery = await this.client.post<WebhookDelivery>(
      API_ENDPOINTS.WEBHOOK_TEST(id),
      {},
      config
    );

    return {
      delivery,
      success: delivery.status === 'success',
      responseTime: delivery.duration || 0,
      statusCode: delivery.responseStatus || 0,
      errorMessage: delivery.errorMessage,
    };
  }

  /**
   * Get webhook statistics
   */
  async getStats(id: number, config?: RequestConfig): Promise<WebhookStats> {
    return this.client.get<WebhookStats>(API_ENDPOINTS.WEBHOOK_STATS(id), config);
  }

  /**
   * Get webhook deliveries
   */
  async getDeliveries(
    id: number,
    pagination?: PaginationParams,
    config?: RequestConfig
  ): Promise<PaginatedResponse<WebhookDelivery>> {
    const params = new URLSearchParams();
    
    if (pagination?.limit) {
      params.set('limit', pagination.limit.toString());
    }
    if (pagination?.offset) {
      params.set('offset', pagination.offset.toString());
    }

    const url = `${API_ENDPOINTS.WEBHOOK_DELIVERIES(id)}${
      params.toString() ? `?${params.toString()}` : ''
    }`;
    
    const response = await this.client.get<{
      deliveries: WebhookDelivery[];
      total: number;
      limit: number;
      offset: number;
    }>(url, config);

    return {
      data: response.deliveries,
      total: response.total,
      limit: response.limit,
      offset: response.offset,
    };
  }

  /**
   * Get failed webhook deliveries
   */
  async getFailedDeliveries(
    id: number,
    limit?: number,
    config?: RequestConfig
  ): Promise<WebhookDelivery[]> {
    const params = new URLSearchParams();
    if (limit) params.set('limit', limit.toString());

    const url = `${API_ENDPOINTS.WEBHOOK_FAILED_DELIVERIES(id)}${
      params.toString() ? `?${params.toString()}` : ''
    }`;
    
    const response = await this.client.get<{ failed_deliveries: WebhookDelivery[] }>(url, config);
    return response.failed_deliveries;
  }

  /**
   * Get a specific webhook delivery
   */
  async getDelivery(deliveryId: number, config?: RequestConfig): Promise<WebhookDelivery> {
    return this.client.get<WebhookDelivery>(API_ENDPOINTS.WEBHOOK_DELIVERY(deliveryId), config);
  }

  /**
   * Retry a failed webhook delivery
   */
  async retryDelivery(deliveryId: number, config?: RequestConfig): Promise<void> {
    await this.client.post(API_ENDPOINTS.WEBHOOK_DELIVERY_RETRY(deliveryId), {}, config);
  }

  /**
   * Get available webhook events
   */
  async getEvents(config?: RequestConfig): Promise<WebhookEventCategories> {
    return this.client.get<WebhookEventCategories>(API_ENDPOINTS.WEBHOOK_EVENTS, config);
  }

  /**
   * Validate webhook URL
   */
  async validateURL(url: string): Promise<{ valid: boolean; error?: string }> {
    try {
      const urlObj = new URL(url);
      
      // Check protocol
      if (!['http:', 'https:'].includes(urlObj.protocol)) {
        return { valid: false, error: 'URL must use HTTP or HTTPS protocol' };
      }
      
      // Check for localhost or private IPs (basic check)
      const hostname = urlObj.hostname.toLowerCase();
      if (hostname === 'localhost' || 
          hostname === '127.0.0.1' || 
          hostname.startsWith('192.168.') ||
          hostname.startsWith('10.') ||
          hostname.startsWith('172.')) {
        return { valid: false, error: 'Webhook URL cannot point to localhost or private IP addresses' };
      }
      
      return { valid: true };
    } catch (error) {
      return { valid: false, error: 'Invalid URL format' };
    }
  }

  /**
   * Generate webhook signature for verification
   */
  generateSignature(payload: string, secret: string): string {
    // This would typically be done server-side, but including for completeness
    // In a real implementation, you'd use a crypto library
    const crypto = require('crypto');
    const hmac = crypto.createHmac('sha256', secret);
    hmac.update(payload);
    return `sha256=${hmac.digest('hex')}`;
  }

  /**
   * Verify webhook signature
   */
  verifySignature(payload: string, signature: string, secret: string): boolean {
    const expectedSignature = this.generateSignature(payload, secret);
    return signature === expectedSignature;
  }

  /**
   * Get webhook health summary
   */
  async getHealthSummary(config?: RequestConfig): Promise<{
    totalWebhooks: number;
    activeWebhooks: number;
    healthyWebhooks: number;
    failingWebhooks: number;
    recentDeliveries: number;
    averageSuccessRate: number;
  }> {
    const webhooks = await this.list({ limit: 1000 }, config);
    
    let activeCount = 0;
    let healthyCount = 0;
    let failingCount = 0;
    let totalSuccessRate = 0;
    let recentDeliveries = 0;

    for (const webhook of webhooks.data) {
      if (webhook.status === 'active') {
        activeCount++;
        
        const successRate = webhook.totalDeliveries > 0 
          ? (webhook.successDeliveries / webhook.totalDeliveries) * 100 
          : 100;
        
        totalSuccessRate += successRate;
        recentDeliveries += webhook.totalDeliveries;
        
        if (successRate >= 95) {
          healthyCount++;
        } else if (successRate < 80) {
          failingCount++;
        }
      }
    }

    return {
      totalWebhooks: webhooks.total,
      activeWebhooks: activeCount,
      healthyWebhooks: healthyCount,
      failingWebhooks: failingCount,
      recentDeliveries,
      averageSuccessRate: activeCount > 0 ? totalSuccessRate / activeCount : 0,
    };
  }

  /**
   * Bulk update webhook status
   */
  async bulkUpdateStatus(
    ids: number[],
    status: 'active' | 'inactive',
    config?: RequestConfig
  ): Promise<void> {
    await this.client.post(
      `${API_ENDPOINTS.WEBHOOKS}/bulk-status`,
      { ids, status },
      config
    );
  }

  /**
   * Bulk delete webhooks
   */
  async bulkDelete(ids: number[], config?: RequestConfig): Promise<void> {
    await this.client.post(
      `${API_ENDPOINTS.WEBHOOKS}/bulk-delete`,
      { ids },
      config
    );
  }

  /**
   * Get webhook templates for common use cases
   */
  async getTemplates(config?: RequestConfig): Promise<{
    templates: Array<{
      name: string;
      description: string;
      events: WebhookEvent[];
      samplePayload: any;
    }>;
  }> {
    return this.client.get(`${API_ENDPOINTS.WEBHOOKS}/templates`, config);
  }

  /**
   * Create webhook from template
   */
  async createFromTemplate(
    templateName: string,
    url: string,
    name: string,
    config?: RequestConfig
  ): Promise<Webhook> {
    return this.client.post<Webhook>(
      `${API_ENDPOINTS.WEBHOOKS}/from-template`,
      { templateName, url, name },
      config
    );
  }

  /**
   * Get delivery insights and recommendations
   */
  async getDeliveryInsights(id: number, config?: RequestConfig): Promise<{
    insights: Array<{
      type: 'error' | 'warning' | 'info';
      message: string;
      recommendation?: string;
    }>;
    performance: {
      averageResponseTime: number;
      successRate: number;
      errorRate: number;
      timeoutRate: number;
    };
    recommendations: string[];
  }> {
    return this.client.get(`${API_ENDPOINTS.WEBHOOK_BY_ID(id)}/insights`, config);
  }
}